### ISSUE TITLE
UnicodeEncodeError in nginxparser.py from RawNginxDumper

### ISSUE THREAD

**Issue by jimhavrilla** [certbot/certbot#8146]
If you're having trouble using Certbot and aren't sure you've found a bug or
request for a new feature, please first try asking for help at
https://community.letsencrypt.org/. There is a much larger community there of
people familiar with the project who will be able to more quickly answer your
questions.

## My operating system is (include version):

CentOS 7, Digital Ocean droplet

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):

certbot-auto

## I ran this command and it produced this output:

certbot-auto

gave correct domain and everything, certificate works, will not install.  Log says there's a unicode error in RawNginxDumper

## Certbot's behavior differed from what I expected because:

It should just work since nginx seems to have no issues...

###### Logs are stored in `/var/log/letsencrypt` by default. Feel free to redact domains, e-mail and IP addresses as you see fit.

```
2020-07-16 20:50:55,377:INFO:certbot_nginx._internal.configurator:Deploying Certificate to VirtualHost /etc/nginx/conf.d/default.conf
2020-07-16 20:50:55,377:DEBUG:certbot.reverter:Creating backup of /etc/nginx/nginx.conf
2020-07-16 20:50:55,378:DEBUG:certbot.reverter:Creating backup of /etc/nginx/conf.d/default.conf
2020-07-16 20:50:55,378:DEBUG:certbot.reverter:Creating backup of /etc/nginx/mime.types
2020-07-16 20:50:55,378:DEBUG:certbot.reverter:Creating backup of /etc/letsencrypt/options-ssl-nginx.conf
2020-07-16 20:50:55,380:DEBUG:certbot._internal.error_handler:Encountered exception:
Traceback (most recent call last):
  File "/opt/eff.org/certbot/venv/lib/python2.7/site-packages/certbot/_internal/client.py", line 521, in deploy_certificate
    self.installer.save()  # needed by the Apache plugin
  File "/opt/eff.org/certbot/venv/lib/python2.7/site-packages/certbot_nginx/_internal/configurator.py", line 1051, in save
    self.parser.filedump(ext='')
  File "/opt/eff.org/certbot/venv/lib/python2.7/site-packages/certbot_nginx/_internal/parser.py", line 250, in filedump
    out = nginxparser.dumps(tree)
  File "/opt/eff.org/certbot/venv/lib/python2.7/site-packages/certbot_nginx/_internal/nginxparser.py", line 140, in dumps
    return str(RawNginxDumper(blocks.spaced))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 662-663: ordinal not in range(128)
```

## Here is the relevant nginx server block or Apache virtualhost for the domain I am configuring:

nginx conf:

server {
    listen 80;
    server_name my-domain.org;

    location / {
        include uwsgi_params;
        uwsgi_pass unix:/user/project/phencards.sock;
    }
}

Certbot also added all the default stuff like `listen 443 ssl` the cert locations etc.

**Comment by alexzorin** [comment]
Could you please dump the raw nginx config which produces this error, and attach it here?

    sudo nginx -T > /tmp/full-nginx.conf

This is/was a Certbot bug where the nginx parser was less tolerant of Unicode than it should be, but I thought it had been fixed since v1.4.0 (#7797).

**Comment by jimhavrilla** [comment]
```
# configuration file /etc/nginx/nginx.conf:

user  nginx;
worker_processes  1;

error_log  /var/log/nginx/error.log warn;
pid        /var/run/nginx.pid;


events {
    worker_connections  1024;
}


http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile        on;
    #tcp_nopush     on;

    keepalive_timeout  65;

    #gzip  on;

    include /etc/nginx/conf.d/*.conf;
}

# configuration file /etc/nginx/mime.types:

types {
    text/html                                        html htm shtml;
    text/css                                         css;
    text/xml                                         xml;
    image/gif                                        gif;
    image/jpeg                                       jpeg jpg;
    application/javascript                           js;
    application/atom+xml                             atom;
    application/rss+xml                              rss;

    text/mathml                                      mml;
    text/plain                                       txt;
    text/vnd.sun.j2me.app-descriptor                 jad;
    text/vnd.wap.wml                                 wml;
    text/x-component                                 htc;

    image/png                                        png;
    image/svg+xml                                    svg svgz;
    image/tiff                                       tif tiff;
    image/vnd.wap.wbmp                               wbmp;
    image/webp                                       webp;
    image/x-icon                                     ico;
    image/x-jng                                      jng;
    image/x-ms-bmp                                   bmp;

    font/woff                                        woff;
    font/woff2                                       woff2;

    application/java-archive                         jar war ear;
    application/json                                 json;
    application/mac-binhex40                         hqx;
    application/msword                               doc;
    application/pdf                                  pdf;
    application/postscript                           ps eps ai;
    application/rtf                                  rtf;
    application/vnd.apple.mpegurl                    m3u8;
    application/vnd.google-earth.kml+xml             kml;
    application/vnd.google-earth.kmz                 kmz;
    application/vnd.ms-excel                         xls;
    application/vnd.ms-fontobject                    eot;
    application/vnd.ms-powerpoint                    ppt;
    application/vnd.oasis.opendocument.graphics      odg;
    application/vnd.oasis.opendocument.presentation  odp;
    application/vnd.oasis.opendocument.spreadsheet   ods;
    application/vnd.oasis.opendocument.text          odt;
    application/vnd.openxmlformats-officedocument.presentationml.presentation
                                                     pptx;
    application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
                                                     xlsx;
    application/vnd.openxmlformats-officedocument.wordprocessingml.document
                                                     docx;
    application/vnd.wap.wmlc                         wmlc;
    application/x-7z-compressed                      7z;
    application/x-cocoa                              cco;
    application/x-java-archive-diff                  jardiff;
    application/x-java-jnlp-file                     jnlp;
    application/x-makeself                           run;
    application/x-perl                               pl pm;
    application/x-pilot                              prc pdb;
    application/x-rar-compressed                     rar;
    application/x-redhat-package-manager             rpm;
    application/x-sea                                sea;
    application/x-shockwave-flash                    swf;
    application/x-stuffit                            sit;
    application/x-tcl                                tcl tk;
    application/x-x509-ca-cert                       der pem crt;
    application/x-xpinstall                          xpi;
    application/xhtml+xml                            xhtml;
    application/xspf+xml                             xspf;
    application/zip                                  zip;

    application/octet-stream                         bin exe dll;
    application/octet-stream                         deb;
    application/octet-stream                         dmg;
    application/octet-stream                         iso img;
    application/octet-stream                         msi msp msm;

    audio/midi                                       mid midi kar;
    audio/mpeg                                       mp3;
    audio/ogg                                        ogg;
    audio/x-m4a                                      m4a;
    audio/x-realaudio                                ra;

    video/3gpp                                       3gpp 3gp;
    video/mp2t                                       ts;
    video/mp4                                        mp4;
    video/mpeg                                       mpeg mpg;
    video/quicktime                                  mov;
    video/webm                                       webm;
    video/x-flv                                      flv;
    video/x-m4v                                      m4v;
    video/x-mng                                      mng;
    video/x-ms-asf                                   asx asf;
    video/x-ms-wmv                                   wmv;
    video/x-msvideo                                  avi;
}

# configuration file /etc/nginx/conf.d/default.conf:
#server {
#    listen       80;
#    server_name  localhost;

    #charset koi8-r;
    #access_log  /var/log/nginx/host.access.log  main;

#    location / {
#        root   /usr/share/nginx/html;
#        index  index.html index.htm;
#    }

server {
    server_name phencards-dev.wglab.org;
    access_log  /var/log/nginx/host.access.log  main;
    set $upstream 157.245.2.226:5005;

    location / {
	proxy_pass_header Authorization;
        proxy_pass  http://$upstream;
	proxy_set_header Host $host;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_http_version 1.1;
	proxy_set_header Connection “”;
	proxy_buffering off;
	client_max_body_size 0;
	proxy_read_timeout 36000s;
	proxy_redirect off;
    }

    #error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

    # proxy the PHP scripts to Apache listening on 127.0.0.1:80
    #
    #location ~ \.php$ {
    #    proxy_pass   http://127.0.0.1;
    #}

    # pass the PHP scripts to FastCGI server listening on 127.0.0.1:9000
    #
    #location ~ \.php$ {
    #    root           html;
    #    fastcgi_pass   127.0.0.1:9000;
    #    fastcgi_index  index.php;
    #    fastcgi_param  SCRIPT_FILENAME  /scripts$fastcgi_script_name;
    #    include        fastcgi_params;
    #}

    # deny access to .htaccess files, if Apache's document root
    # concurs with nginx's one
    #
    #location ~ /\.ht {
    #    deny  all;
    #}

    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/phencards-dev.wglab.org/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/phencards-dev.wglab.org/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot


}



server {
    if ($host = phencards-dev.wglab.org) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


    listen 80;
    server_name phencards-dev.wglab.org;
    return 404; # managed by Certbot


}

# configuration file /etc/letsencrypt/options-ssl-nginx.conf:
# This file contains important security parameters. If you modify this file
# manually, Certbot will be unable to automatically provide future security
# updates. Instead, Certbot will print and log an error message with a path to
# the up-to-date file that you will need to refer to when manually updating
# this file.

ssl_session_cache shared:le_nginx_SSL:10m;
ssl_session_timeout 1440m;

ssl_protocols TLSv1.2 TLSv1.3;
ssl_prefer_server_ciphers off;

ssl_ciphers "ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384";


```

**Comment by alexzorin** [comment]
Well, the Unicode in that file which is likely the cause of the error is:

>	proxy_set_header Connection “”;

Those quotes are U+201C and U+201D, rather than the usual ASCII ones. Possibly you copied this from the Let's Encrypt community forum, where the forum software converts them unlesss they're in code blocks.

Nonetheless, Certbot should not crash on this, because exactly this behavior has already been fixed in the past. I can't reproduce it on latest `certbot-auto`. Your `certbot-auto` is not being held at a previous version, is it?

    sudo certbot-auto --version

**Comment by jimhavrilla** [comment]
Yep that was it, thanks.  I feel dumb.  Nice work.

version 1.6.0 btw.

**Comment by alexzorin** [comment]
Actually, I managed to reproduce this on CentOS 7 + `certbot-auto` 1.6.0. 

It might be a Python 2 vs Python 3 problem.

**Comment by jimhavrilla** [comment]
You were right.  I changed the quotes and it no longer threw an error.

**Comment by ynikitenko** [comment]
I have a similar error,

    UnicodeEncodeError: 'ascii' codec can't encode characters in position 436-451: ordinal not in range(128)

when running 

    certbot --nginx

on CentOS 7. Surprisingly, that is a recent package (1.9.0). 
It prohibits any non-ascii data in nginx configuration file for my site. However, my site uses unicode folders for media files, so I can't simply change the domain name to punicode. I delete unicode, run certbot, and then restore the proper paths.

Apart from that, everything is great, thanks a lot for the program!

**Comment by ynikitenko** [comment]
However, it turns out that I can't automatically renew this certificate any longer.

    certbot renew --dry-run

(from the log file):

    File "/usr/lib/python2.7/site-packages/certbot_nginx/_internal/nginxparser.py", line 140, in dumps
        return str(RawNginxDumper(blocks.spaced))
    UnicodeEncodeError: 'ascii' codec can't encode characters in position 587-591: ordinal not in range(128)

Surprisingly, it parses nginx even when it is already installed. Is it really necessary? Can't it just renew the installed certificates? When I run

    certbot certonly --nginx --dry-run

it gives the same UnicodeError. Shouldn't it just install the certificate to its own path?

**Comment by alexzorin** [comment]
>Is it really necessary?

If you use `--nginx` to renew the certificate, then it needs to be able to load, modify and save your configuration. As a result, the `UnicodeError` will be encountered during renewal.

One possible workaround would be to use `--webroot` rather than `--nginx` to renew.

Another workaround would be to [install Certbot from snap](https://certbot.eff.org/lets-encrypt/centosrhel7-nginx) rather than from CentOS EPEL. This should work because the snap uses Python 3, where (I believe) this issue doesn't exist.

**Comment by ynikitenko** [comment]
" to use --webroot"  - thanks, I'll try that. Initially I tried to use snap, but it failed because I had no squashfs/fuse on my OpenVZ VPS. So I may just easier write a custom script to apply changes back and forward.

However, I'm a bit surprised that Python 2 is the major language for this bot (at least for OS packages). I very much hope that it will be fixed in Python 2 - do you think it might be done in several months, or will the command just drop Python 2 support (and these OS packages)?

**Comment by alexzorin** [comment]
I previously asked the team before whether this should be fixed, and the feedback was that fixing Python 2-related things is not a priority due to its imminent deprecation.

I honestly don't know whether/when CentOS EPEL7's Certbot packages will move to Python 3.

I have proposed a fix in #8480 with the hope that it can be released in EPEL7 as Certbot 1.10, but it could be too late.
